File: filesystem/lisp/GWOS_lib/libc.c
#include "libc.h"

int strcmp(const char *s1, const char *s2){
	while(*s1 && (*s1 == *s2)){
		s1++;
		s2++;
	}
	return *(const unsigned char *)s1 - *(const unsigned char*)s2;
}

struct WINDOW *window_open(char *name, uint32_t copyOnPrompt){
	register unsigned int eax asm("eax");
    register unsigned int ebx asm("ebx");
    register unsigned int ecx asm("ecx");
    register unsigned int edx asm("edx"); 
	ecx = (unsigned int) copyOnPrompt;   
    ebx = (unsigned int) name;
    eax = 0x01;
    asm("int 0x80");
    struct WINDOW *win = (void *) eax;
	return win;
}

void window_close(struct WINDOW *window){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	eax = 0x02;
	ebx = (uint32_t) window;
	asm("int 0x80");
}

void window_update(){
	register uint32_t eax asm("eax");
	eax = 0x03;
	asm("int 0x80");
}

void drawChar(uint32_t x, uint32_t y, char c){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	register uint32_t ecx asm("ecx");
	register uint32_t edx asm("edx");
	edx = c;
	ecx = y;
	ebx = x;
	eax = 0x04;
	asm("int 0x80");
}

char getc(){
	
	char (*getc_blk)(void);
	register uint32_t eax asm("eax");
	eax = 0x05;
	asm("int 0x80");
	getc_blk = (char (*)(void)) eax;
	//return ((char (*)(void))eax)();
	return getc_blk();
	/*
	uint32_t c = 0;
	while(!c){
		register uint32_t eax asm("eax");
		eax = 0x05;
		asm("int 0x80");
		c = eax;
	}
	return (char) c;
	*/
}

void exec(char *filename, int argc, char **argv){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	register uint32_t ecx asm("ecx");
	register uint32_t edx asm("edx");
	edx = (uint32_t) argv;
	ecx = (uint32_t) argc;
	ebx = (uint32_t) filename;
	eax = 0x06;
	asm("int 0x80");
}
void set_schedule(ScheduleType type){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = type;
	eax = 0x07;
	asm("int 0x80");
}

void print(char *str){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) str;
	eax = 0x08;
	asm("int 0x80");
}

void print_arg(char *str, uint32_t arg){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	register uint32_t ecx asm("ecx");
	ecx = (uint32_t) arg;
	ebx = (uint32_t) str;
	eax = 0x0C;
	asm("int 0x80");
}

void print_float(char *str, float arg){
	print_arg(str, *((uint32_t *) &arg));
}

struct CONSOLE *console_open(){
	register uint32_t eax asm("eax");
	eax = 0x09;
	asm("int 0x80");
	struct CONSOLE *console = (struct CONSOLE *) eax;
	return console;
}

void console_close(){
	register uint32_t eax asm("eax");
	eax = 0x0A;
	asm("int 0x80");
}

void *kmalloc(uint32_t size){
	register uint32_t eax asm("eax");
	eax = 0x0B;
	asm("int 0x80");
	return (void *) eax;
}

struct FILE *fopen(char *file){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) file;
	eax = 0x0D;
	asm("int 0x80");
	struct FILE *f = (struct FILE *) eax;
	return f;
}

void fclose(struct FILE *file){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) file;
	eax = 0x0E;
	asm("int 0x80");
}

char fgetc(struct FILE *file){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) file;
	eax = 0x0F;
	asm("int 0x80");
	return (char) eax;
}

int fsize(struct FILE *file){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) file;
	eax = 0x10;
	asm("int 0x80");
	return (int) eax;
}

int fcopy(struct FILE *file, char *buf, int buf_size){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	register uint32_t ecx asm("ecx");
	register uint32_t edx asm("edx");
	edx = buf_size;
	ecx = (uint32_t) buf;
	ebx = (uint32_t) file;
	eax = 0x11;
	asm("int 0x80");
}

int changeDirectory(char *change){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) change;
	eax = 0x12;
	asm("int 0x80");
	return eax;
}

char *getDirectory(){
	register uint32_t eax asm("eax");
	eax = 0x13;
	asm("int 0x80");
	return (char *) eax;
}

void printDirectoryContents(char *path){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) path;
	eax = 0x14;
	asm("int 0x80");
}

int fexists(char *path){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) path;
	eax = 0x15;
	asm("int 0x80");
	return eax;
}

int fextend(struct FILE *file, uint32_t amount){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	register uint32_t ecx asm("ecx");
	ecx = amount;
	ebx = (uint32_t) file;
	eax = 0x22;
	asm("int 0x80");
	return eax;
}

struct FEATURE_INFO getKernelFeature(KERNEL_FEATURE feature){
	void *addr;
	uint32_t size;
	struct FEATURE_INFO info;
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) feature;
	eax = 0x17;
	asm("int 0x80");
	addr = (void *) eax;
	size = ebx;
	info.addr = addr;
	info.size = size;
	return info;
}

void fseek(struct FILE *file, int idx){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	register uint32_t ecx asm("ecx");
	ecx = (uint32_t) idx;
	ebx = (uint32_t) file;
	eax = 0x18;
	asm("int 0x80");
}

void fputc(struct FILE *file, char c){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	register uint32_t ecx asm("ecx");
	ecx = (uint32_t) c;
	ebx = (uint32_t) file;
	eax = 0x19;
	asm("int 0x80");	
}

void fmkdir(char *path){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) path;
	eax = 0x1A;
	asm("int 0x80");
}

void fmkfile(char *path, int size){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	register uint32_t ecx asm("ecx");
	ecx = (uint32_t) size;
	ebx = (uint32_t) path;
	eax = (uint32_t) 0x1B;
	asm("int 0x80");
}

int getArrowKey(){
	register uint32_t eax asm("eax");
	eax = 0x1C;
	asm("int 0x80");
	return eax;
}

void srand(uint32_t seed){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = seed;
	eax = 0x1D;
	asm("int 0x80");
}

int rand(){
	register uint32_t eax asm("eax");
	eax = 0x1E;
	asm("int 0x80");
	return eax;
}

void exit(int code){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) code;
	eax = 0x1F;
	asm("int 0x80");
	return;
}

void *requestRegion(unsigned int bytes){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = bytes;
	eax = 0x20;
	asm("int 0x80");
	return (void *) eax;
}

void addKbdEventHandler(void (*handler)(unsigned char)){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) handler;
	eax = 0x21;
	asm("int 0x80");
}

void *requestRegionAt(unsigned int bytes, unsigned int addr){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	register uint32_t ecx asm("ecx");
	ecx = addr;
	ebx = bytes;
	eax = 0x23;
	asm("int 0x80");
	return (void *) eax;
}

void attachTimerCallback(unsigned int timer, void *callback){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	register uint32_t ecx asm("ecx");
	ecx = (uint32_t) callback;
	ebx = (uint32_t) timer;
	eax = 0x24;
	asm("int 0x80");
}

void dprint(char *msg){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) msg;
	eax = 0x25;
	asm("int 0x80");
}

void start_manual_task(void *addr, char *name){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	register uint32_t ecx asm("ecx");
	ecx = (uint32_t) name;
	ebx = (uint32_t) addr;
	eax = 0x26;
	asm("int 0x80");
}

struct MouseStatus getMouse(){
	struct MouseStatus *mousePtr;
	struct MouseStatus mouse;
	register uint32_t eax asm("eax");
	eax = 0x27;
	asm("int 0x80");
	mousePtr = (struct MouseStatus *) eax;
	mouse = *mousePtr;
	return mouse;
}

struct PCSpeaker_Handle *getPCSpeaker(){
	struct PCSpeaker_Handle *handle;
	register uint32_t eax asm("eax");
	eax = 0x28;
	asm("int 0x80");
	handle = (struct PCSpeaker_Handle *) eax;
	return handle;
}

uint32_t *getTimerTickHandle(){
	uint32_t *tick_handle;
	register uint32_t eax asm("eax");
	eax = 0x29;
	asm("int 0x80");
	tick_handle = (uint32_t *) eax;
	return tick_handle;
}

void addMouseEventHandler(void (*handler)(void)){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) handler;
	eax = 0x2A;
	asm("int 0x80");
}

void freeRegion(void *addr, uint32_t size){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	register uint32_t ecx asm("ecx");
	ecx = size;
	ebx = (uint32_t) addr;
	eax = 0x2B;
	asm("int 0x80");
}

void addEndCallback(void (*handler)(void)){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) handler;
	eax = 0x2C;
	asm("int 0x80");
}

struct ViewportFunctions *viewport_get_funcs(){
	register uint32_t eax asm("eax");
	eax = 0x2D;
	asm("int 0x80");
	return (struct ViewportFunctions *) eax;
}

struct CONSOLE *console_open_vp(struct Viewport *vp){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) vp;
	eax = 0x2E;
	asm("int 0x80");
	struct CONSOLE *console = (struct CONSOLE *) eax;
	return console;
}

struct RealTimeClock *get_rtc(){
	register uint32_t eax asm("eax");
	eax = 0x2F;
	asm("int 0x80");
	return (struct RealTimeClock *) eax;
}

struct DirectoryListing getDirectoryListing(char *path){
	struct DirectoryListing (*listFunc)(char *);
	register uint32_t eax asm("eax");
	eax = 0x30;
	asm("int 0x80");
	listFunc = (struct DirectoryListing (*)(char *)) eax;
	return listFunc(path);
}

void print_serial(char *str){
	register uint32_t eax asm("eax");
	register uint32_t ebx asm("ebx");
	ebx = (uint32_t) str;
	eax = 0x31;
	asm("int 0x80");
}

File: filesystem/utils/editor/libc.zig
pub const @"bool" = u8;
pub const struct_IVec2 = extern struct {
    x: c_int = @import("std").mem.zeroes(c_int),
    y: c_int = @import("std").mem.zeroes(c_int),
};
// ../../OS/libc/libc.h:25:13: warning: struct demoted to opaque type - has bitfield
pub const struct_MouseButtons = opaque {};
pub const struct_MouseStatus = extern struct {
    pos: struct_IVec2 = @import("std").mem.zeroes(struct_IVec2),
    buttons: struct_MouseButtons = @import("std").mem.zeroes(struct_MouseButtons),
    lastDelta: struct_IVec2 = @import("std").mem.zeroes(struct_IVec2),
};
pub const struct_PCSpeaker_Handle = extern struct {
    play: ?*const fn (u32) callconv(.C) void = @import("std").mem.zeroes(?*const fn (u32) callconv(.C) void),
    mute: ?*const fn () callconv(.C) void = @import("std").mem.zeroes(?*const fn () callconv(.C) void),
    beep: ?*const fn () callconv(.C) void = @import("std").mem.zeroes(?*const fn () callconv(.C) void),
};
pub const struct_WINDOW = extern struct {
    backbuffer: [*c]u32 = @import("std").mem.zeroes([*c]u32),
    name: [21]u8 = @import("std").mem.zeroes([21]u8),
    active: @"bool" = @import("std").mem.zeroes(@"bool"),
    width: u32 = @import("std").mem.zeroes(u32),
    height: u32 = @import("std").mem.zeroes(u32),
    copyOnPromptOnly: @"bool" = @import("std").mem.zeroes(@"bool"),
};
pub const ALWAYS: c_int = 1;
pub const ONFOCUS: c_int = 2;
pub const NOCHILD: c_int = 3;
pub const NEVER: c_int = 4;
pub const ScheduleType = c_uint;
const struct_unnamed_1 = extern struct {
    fg: u32 = @import("std").mem.zeroes(u32),
    bg: u32 = @import("std").mem.zeroes(u32),
};
pub const struct_CONSOLE = extern struct {
    window: [*c]struct_WINDOW = @import("std").mem.zeroes([*c]struct_WINDOW),
    width: u32 = @import("std").mem.zeroes(u32),
    height: u32 = @import("std").mem.zeroes(u32),
    buf_size: u32 = @import("std").mem.zeroes(u32),
    buf: [*c]u8 = @import("std").mem.zeroes([*c]u8),
    active: @"bool" = @import("std").mem.zeroes(@"bool"),
    last_cursor: c_int = @import("std").mem.zeroes(c_int),
    cursor: c_int = @import("std").mem.zeroes(c_int),
    Line: u32 = @import("std").mem.zeroes(u32),
    Start: u32 = @import("std").mem.zeroes(u32),
    LinePlace: u32 = @import("std").mem.zeroes(u32),
    LastLine: u32 = @import("std").mem.zeroes(u32),
    color: struct_unnamed_1 = @import("std").mem.zeroes(struct_unnamed_1),
};
pub const struct_File_Info = extern struct {
    drive: ?*anyopaque = @import("std").mem.zeroes(?*anyopaque),
    sector: u32 = @import("std").mem.zeroes(u32),
    size: u32 = @import("std").mem.zeroes(u32),
};
pub const struct_FILE = extern struct {
    info: struct_File_Info = @import("std").mem.zeroes(struct_File_Info),
    head: c_int = @import("std").mem.zeroes(c_int),
};
pub extern fn strcmp(s1: [*c]const u8, s2: [*c]const u8) c_int;
pub extern fn window_open(name: [*c]u8, copyOnPrompt: u32) [*c]struct_WINDOW;
pub extern fn window_close(window: [*c]struct_WINDOW) void;
pub extern fn window_update(...) void;
pub extern fn drawChar(x: u32, y: u32, c: u8) void;
pub extern fn getc(...) u8;
pub extern fn exec(filename: [*c]u8, argc: c_int, argv: [*c][*c]u8) void;
pub extern fn set_schedule(@"type": ScheduleType) void;
pub extern fn print(str: [*c]u8) void;
pub extern fn print_arg(str: [*c]u8, arg: u32) void;
pub extern fn print_float(str: [*c]u8, arg: f32) void;
pub extern fn console_open(...) [*c]struct_CONSOLE;
pub extern fn console_close(...) void;
pub extern fn kmalloc(size: u32) ?*anyopaque;
pub extern fn fopen(file: [*c]u8) [*c]struct_FILE;
pub extern fn fclose(file: [*c]struct_FILE) void;
pub extern fn fgetc(file: [*c]struct_FILE) u8;
pub extern fn fsize(file: [*c]struct_FILE) c_int;
pub extern fn fcopy(file: [*c]struct_FILE, buf: [*c]u8, buf_size: c_int) c_int;
pub extern fn changeDirectory(change: [*c]u8) c_int;
pub extern fn getDirectory(...) [*c]u8;
pub extern fn printDirectoryContents(path: [*c]u8) void;
pub extern fn fexists(path: [*c]u8) c_int;
pub const FEAT_FONT: c_int = 1;
pub const FEAT_KEYBOARD: c_int = 2;
pub const FEAT_PAGETABLE: c_int = 3;
pub const FEAT_TASKTABLE: c_int = 4;
pub const FEAT_PROGRAMBASE: c_int = 5;
pub const FEAT_FRAMEBUFFER: c_int = 6;
pub const FEAT_TIMER: c_int = 7;
pub const FEAT_KEYPRESSMAP: c_int = 8;
pub const KERNEL_FEATURE = c_uint;
pub const struct_FEATURE_INFO = extern struct {
    addr: ?*anyopaque = @import("std").mem.zeroes(?*anyopaque),
    size: u32 = @import("std").mem.zeroes(u32),
};
pub extern fn getKernelFeature(feature: KERNEL_FEATURE) struct_FEATURE_INFO;
pub extern fn fseek(file: [*c]struct_FILE, idx: c_int) void;
pub extern fn fputc(file: [*c]struct_FILE, c: u8) void;
pub extern fn fmkdir(path: [*c]u8) void;
pub extern fn fmkfile(path: [*c]u8, size: c_int) void;
pub extern fn fextend(file: [*c]struct_FILE, amount: u32) c_int;
pub extern fn getArrowKey(...) c_int;
pub extern fn srand(seed: u32) void;
pub extern fn rand(...) c_int;
pub extern fn exit(code: c_int) noreturn;
pub extern fn requestRegion(bytes: c_uint) ?*anyopaque;
pub extern fn addKbdEventHandler(handler: ?*const fn (u8) callconv(.C) void) void;
pub extern fn requestRegionAt(bytes: c_uint, addr: c_uint) ?*anyopaque;
pub extern fn attachTimerCallback(timer: c_uint, callback: ?*anyopaque) void;
pub extern fn dprint(msg: [*c]u8) void;
pub extern fn start_manual_task(addr: ?*anyopaque, name: [*c]u8) void;
pub extern fn getMouse(...) ?*struct_MouseStatus;
pub extern fn getPCSpeaker(...) [*c]struct_PCSpeaker_Handle;
pub extern fn getTimerTickHandle(...) [*c]u32;
pub extern fn addMouseEventHandler(handler: ?*const fn () callconv(.C) void) void;
pub extern fn freeRegion(addr: ?*anyopaque, size: u32) void;
pub extern fn addEndCallback(handler: ?*const fn () callconv(.C) void) void;
pub const struct_Location = extern struct {
    x: c_int = @import("std").mem.zeroes(c_int),
    y: c_int = @import("std").mem.zeroes(c_int),
    w: c_int = @import("std").mem.zeroes(c_int),
    h: c_int = @import("std").mem.zeroes(c_int),
};
pub const VP_FOCUSED: c_int = 0;
pub const VP_UNFOCUSED: c_int = 1;
pub const VP_KEY: c_int = 2;
pub const VP_EXIT: c_int = 3;
pub const VP_MINIMIZE: c_int = 4;
pub const VP_MAXIMIZE: c_int = 5;
pub const VP_CLICK: c_int = 6;
pub const VIEWPORT_EVENT_TYPE = c_uint;
pub const struct_Viewport = extern struct {
    loc: struct_Location = @import("std").mem.zeroes(struct_Location),
    oldLoc: struct_Location = @import("std").mem.zeroes(struct_Location),
    frontbuf: [*c]u32 = @import("std").mem.zeroes([*c]u32),
    backbuf: [*c]u32 = @import("std").mem.zeroes([*c]u32),
    buf_size: u32 = @import("std").mem.zeroes(u32),
    resizeable: u8 = @import("std").mem.zeroes(u8),
    minimized: @"bool" = @import("std").mem.zeroes(@"bool"),
    open: @"bool" = @import("std").mem.zeroes(@"bool"),
    minimized_w: c_int = @import("std").mem.zeroes(c_int),
    minimized_h: c_int = @import("std").mem.zeroes(c_int),
    title: [*c]u8 = @import("std").mem.zeroes([*c]u8),
    owner_program_slot: c_int = @import("std").mem.zeroes(c_int),
    owner_task_id: c_int = @import("std").mem.zeroes(c_int),
    ascii: u8 = @import("std").mem.zeroes(u8),
    event_handler: ?*const fn ([*c]struct_Viewport, VIEWPORT_EVENT_TYPE) callconv(.C) void = @import("std").mem.zeroes(?*const fn ([*c]struct_Viewport, VIEWPORT_EVENT_TYPE) callconv(.C) void),
    click_events_enabled: @"bool" = @import("std").mem.zeroes(@"bool"),
};
pub const struct_ViewportFunctions = extern struct {
    open: ?*const fn (c_int, c_int, [*c]u8) callconv(.C) [*c]struct_Viewport = @import("std").mem.zeroes(?*const fn (c_int, c_int, [*c]u8) callconv(.C) [*c]struct_Viewport),
    close: ?*const fn ([*c]struct_Viewport) callconv(.C) void = @import("std").mem.zeroes(?*const fn ([*c]struct_Viewport) callconv(.C) void),
    set_buffer: ?*const fn ([*c]struct_Viewport, [*c]u32, u32) callconv(.C) void = @import("std").mem.zeroes(?*const fn ([*c]struct_Viewport, [*c]u32, u32) callconv(.C) void),
    copy: ?*const fn ([*c]struct_Viewport) callconv(.C) void = @import("std").mem.zeroes(?*const fn ([*c]struct_Viewport) callconv(.C) void),
    add_event_handler: ?*const fn ([*c]struct_Viewport, ?*const fn ([*c]struct_Viewport, VIEWPORT_EVENT_TYPE) callconv(.C) void) callconv(.C) void = @import("std").mem.zeroes(?*const fn ([*c]struct_Viewport, ?*const fn ([*c]struct_Viewport, VIEWPORT_EVENT_TYPE) callconv(.C) void) callconv(.C) void),
    drawChar: ?*const fn ([*c]struct_Viewport, c_int, c_int, u8, u32, u32) callconv(.C) void = @import("std").mem.zeroes(?*const fn ([*c]struct_Viewport, c_int, c_int, u8, u32, u32) callconv(.C) void),
    getc: ?*const fn ([*c]struct_Viewport) callconv(.C) u8 = @import("std").mem.zeroes(?*const fn ([*c]struct_Viewport) callconv(.C) u8),
};
pub extern fn viewport_get_funcs(...) [*c]struct_ViewportFunctions;
pub extern fn console_open_vp(vp: [*c]struct_Viewport) [*c]struct_CONSOLE;
pub const struct_RealTimeClock = extern struct {
    second: u8 = @import("std").mem.zeroes(u8),
    minute: u8 = @import("std").mem.zeroes(u8),
    hour: u8 = @import("std").mem.zeroes(u8),
    day: u8 = @import("std").mem.zeroes(u8),
    month: u8 = @import("std").mem.zeroes(u8),
    year: c_uint = @import("std").mem.zeroes(c_uint),
    msec: c_uint = @import("std").mem.zeroes(c_uint),
};
pub extern fn get_rtc(...) [*c]struct_RealTimeClock;
pub const struct_DirectoryEntry = extern struct {
    filename: [50]u8 = @import("std").mem.zeroes([50]u8),
    name_len: c_int = @import("std").mem.zeroes(c_int),
    type: u32 = @import("std").mem.zeroes(u32),
};
pub const struct_DirectoryListing = extern struct {
    directory_path: [*c]u8 = @import("std").mem.zeroes([*c]u8),
    directory_path_len: c_int = @import("std").mem.zeroes(c_int),
    entries: [*c]struct_DirectoryEntry = @import("std").mem.zeroes([*c]struct_DirectoryEntry),
    num_entries: c_int = @import("std").mem.zeroes(c_int),
};
pub extern fn getDirectoryListing(path: [*c]u8) struct_DirectoryListing;
pub extern fn print_serial(str: [*c]u8) void;
pub extern fn task_lock(state: c_int) void;
pub extern fn write_serial(c: u8) void;
pub extern fn yield(...) void;
pub const @"true" = @as(c_int, 1);
pub const @"false" = @as(c_int, 0);
pub const NULL = @as(c_int, 0);
pub const IVec2 = struct_IVec2;
pub const MouseButtons = struct_MouseButtons;
pub const MouseStatus = struct_MouseStatus;
pub const PCSpeaker_Handle = struct_PCSpeaker_Handle;
pub const WINDOW = struct_WINDOW;
pub const CONSOLE = struct_CONSOLE;
pub const File_Info = struct_File_Info;
pub const FILE = struct_FILE;
pub const FEATURE_INFO = struct_FEATURE_INFO;
pub const Location = struct_Location;
pub const Viewport = struct_Viewport;
pub const ViewportFunctions = struct_ViewportFunctions;
pub const RealTimeClock = struct_RealTimeClock;
pub const DirectoryEntry = struct_DirectoryEntry;
pub const DirectoryListing = struct_DirectoryListing;
